### PR TITLE
When a sender version is less than 5.0 do not check the message intent

### LINKED ISSUE #67: Request and response between callbacks and NServiceBus v4 throws an InvalidOperationException
## Who's affected

* Customers using Request and Response between callbacks and NServiceBus v4 with object messages

## Symptoms

When an NServiceBus v6 sender/receiver issues a request/response with the callbacks to an NServiceBus v4 receiver which replies with an object message (`bus.Reply(new ObjectResponseMessage())`) an `InvalidOperationException` is thrown in the NServiceBus v6 receiver indicating that no handler could be found

```
System.InvalidOperationException: No handlers could be found for message type: ObjectResponseMessage
```

More details can be found https://github.com/Particular/NServiceBus.Callbacks/issues/66#issuecomment-312695212

## Expected behavior

The callback registered should be triggered

## Details

NServiceBus v4 issues a reply with `MessageIntentEnum.Reply` which then triggers the [version check](https://github.com/Particular/NServiceBus.Callbacks/blob/develop/src/NServiceBus.Callbacks/IncomingContextExtensions.cs#L34). 

1. v6 sends a message to 4.7.x
1. 4.7.x replies with a message and seems to set `MessageIntentEnum.Reply` but what goes out is `MessageIntentEnum.Send`
image
image
1. Because the message arrives with intent Send the following code trips over
https://github.com/Particular/NServiceBus.Callbacks/blob/develop/src/NServiceBus.Callbacks/IncomingContextExtensions.cs#L34
since with check for Min supported version to be 4.3 https://github.com/Particular/NServiceBus.Callbacks/blob/develop/src/NServiceBus.Callbacks/IncomingContextExtensions.cs#L49

## Solution

When the version that replied is less or equal to 4.7 do not check the message intent

### PR DESCRIPTION
**Pull request by danielmarbach**
Fixes #67 

[More details](https://github.com/Particular/NServiceBus.Callbacks/issues/66#issuecomment-312797501)

### CODE REVIEW

**Comment by subas4u** [comment]
Hi All, Thank you for your quick response. Do you know when this hot-fix will be available via NuGet package ("NServiceBus.Callbacks")? I'm the one one who raised this issue via Google Group https://groups.google.com/forum/#!topic/particularsoftware/551q9DNTBCQ.

File: src/NServiceBus.Callbacks/IncomingContextExtensions.cs [hunk 1]
```diff
@@ -46,6 +46,6 @@ static string GetCorrelationId(this IMessageProcessingContext context)
             return context.MessageHeaders.TryGetValue(Headers.CorrelationId, out str) ? str : null;
         }
 
-        static Version minimumVersionThatSupportMessageIntent_Reply = new Version(4, 3);
+        static Version minimumVersionThatSupportMessageIntent_Reply = new Version(4, 7);
```
Comment (SimonCropp) [inline src/NServiceBus.Callbacks/IncomingContextExtensions.cs]: why do we hard code the minor here? why  not "less than 5"?

File: src/NServiceBus.Callbacks/IncomingContextExtensions.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ static class IncomingContextExtensions
                 Version parsedVersion;
                 if (Version.TryParse(version, out parsedVersion))
                 {
-                    if (parsedVersion < minimumVersionThatSupportMessageIntent_Reply)
+                    if (parsedVersion <= minimumVersionThatSupportMessageIntent_Reply)
```
Comment (SimonCropp) [inline src/NServiceBus.Callbacks/IncomingContextExtensions.cs]: `parsedVersion` is `{4.7.12}` so this wont evaluate to true

**Comment by SimonCropp** [comment]
so changing 

```
static Version minimumVersionThatSupportMessageIntent_Reply = new Version(4, 7);
```
to 
```
static Version minimumVersionThatSupportMessageIntent_Reply = new Version(5, 0);
```

and this
```
if (parsedVersion <= minimumVersionThatSupportMessageIntent_Reply)
```
back to this
```
if (parsedVersion < minimumVersionThatSupportMessageIntent_Reply)
```

made it pass the wire compat tests

**Comment by SimonCropp** [comment]
and the same problem exists in callbacks v1 and v1.1

**Comment by danielmarbach** [comment]
I verified this with your repro and it worked. I can do more digging today.
The idea was that we keep it aligned with the previous code. The problem is
that 4.3 introduced this intent but 4.7.x didn't use it correctly. We could
in theorry change it to major 5. correct. V1 is out of support, v1.1 will
be patched (there is an issue raised).

> --

Have you tried NServiceBus V6 <https://particular.net/nservicebus-v6> yet?

Learn how to scale applications with microservices and NServiceBus 6
<https://www.pluralsight.com/courses/microservices-nservicebus6-scaling-applications>
.
------------------------------

*Join Udi Dahan's next Advanced Distributed Design Course:*

September 2017 Los Angeles, CA, USA <http://particular.net/adsd>

November 2017 Sydney, Australia <http://particular.net/adsd>
------------------------------

**Comment by timbussmann** [comment]
@SimonCropp changes seem to make sense 👍 I didn't like the `< 4.8` attempt as 4.8 doesn't exist, but `< 5` seems to be fine in that regards.

**Comment by SimonCropp** [comment]
@timbussmann @danielmarbach BTW the wire compat tests are great at telling u something is wrong, but horrible at debugging specific scenarios. so i did some ad hoc scripting and extracted this https://github.com/SimonCropp/scratch/tree/master/AllCallbacks  so it is just the scenario in question and almost all the vectors. i didnt bother with all the combos for callbacks 1.x given the code base is so similar to v2.

**Comment by danielmarbach** [comment]
@subas4u we are in the process of releasing the changes

**Comment by subas4u** [comment]
Thanks all. The hot fix solved my issue.

**Pull request merged**